Updated mongodb nodes : Merged IN and OUT nodes

mickev36 opened the pull request:
+ every operation emits an output with the query result or an error object

Discussed in https://discourse.nodered.org/t/mongodb-node-improvements/4442 and on slack with @knolleary a couple weeks ago, which proposed to merge my fork.

I had my js formatter enabled, so you might have some troube with the diff. If you cannot accept it, please let me know, I will re-do it without the code formatter.

<!--
## Before you hit that Submit button....

Please read our [contribution guidelines](https://github.com/node-red/node-red-nodes/blob/master/CONTRIBUTING.md)
before submitting a pull-request.

## Types of changes

What types of changes does your code introduce?
Put an `x` in the boxes that apply
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

<!--
If you want to raise a pull-request with a new feature, or a refactoring
of existing code, it **may well get rejected** if it hasn't been discussed on
the [mailing list](https://groups.google.com/forum/#!forum/node-red) or
[slack team](https://nodered.org/slack) first.

-->

## Proposed changes

<!-- Describe the nature of this change. What problem does it address? -->

I think the nodes for node-red-node-mongodb (which is provided with the extra nodes package) should provide always provide an output, so we can process the status of the query (Did it fail ? Did it succeed ? When ?)

Done in this PR :
Merges the two mongodb nodes. Every action outputs a status (query status or error).

## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red-nodes/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the mailing list/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

knolleary commented:
Hi,

we need to have more discussion about your proposal.

I see neither @dceejay or myself responded to the discussion on the forum last October and I don't recall having any in depth conversation on slack about it - apologies if I'm misremembering. If we had we would have pointed out that merging existing nodes would be a massively breaking change that we wouldn't consider as the starting point for any change.

A much more modest change would be to add the insert/save/update/remove options to the existing MongoDB In node and leave the Out node largely untouched (albeit with a bit of internal refactoring to reduce code duplication).

That would satisfy the requirement of being able to do a write to the database *and* get a result back - whilst maintaining 100% backward compatibility.

Apologies for missing the original forum discussion - its a worthwhile requirement to satisfy, we just need to find the right approach.

mickev36 commented:
Hi,
Thanks for your answer. 

I understand that introducing this breaking change is not possible. 
What you describe seems great, i'm willing to implement it (add missing operations to the IN node to output stuff, while avoiding having code duplicated between the two nodes).

I'm not sure if I should open up another PR since the name won't be matching now ? 
Thanks

mickev36 commented:
Hi @knolleary @dceejay 

I've updated the PR with what we discussed : I added the 4 missing operations (insert/save/update/delete) to the IN node, and emitted the mongodb operation result. I did a simple factorization so that no code is duplicated.

Let me know if there's anything more 😄

mickev36 commented:
Hi @knolleary @dceejay 

Any update on the subject ? :-)

Cheers

duartegarin commented:
Hi guys,
Also facing the exact same issue.
Any updates on this?

mickev36 commented:
> Hi guys,
> Also facing the exact same issue.
> Any updates on this?

Hello duartegarin,

In the meantime, i've made a npm package that fixes this issue : https://www.npmjs.com/package/node-red-contrib-mongodb

The change should be transparent when the PR is merged.
Feel free to message me if you encounter any issue. 
Have a good day

mickev36 commented:
> Hi guys,
> Also facing the exact same issue.
> Any updates on this?

Hello @duartegarin,

In the meantime, i've made a npm package that fixes this issue : https://www.npmjs.com/package/node-red-contrib-mongodb

The change should be transparent when the PR is merged.
Feel free to message me if you encounter any issue.
Have a good day

albfan commented:
Let me know if I can collaborate onn this PR. get notified when a mongo operation ends is useful for flows where mongodb input  is needed.

mickev36 commented:
Hi @knolleary 

Is there any updates on this MR ? I see it's been removed from "To Do", is there anything that needs to be done ?

Thanks in advance !

ruralcoder commented:
Any updates on this?

mickev36 commented:
> 
> 
> Any updates on this? Nodes are useless unless you get feedback or the newly created _id.

Hello ruralcoder,

Unfortunately, I haven't heard back from the core devs in over a year and a half, so i'm guessing this will never be merged.
@knolleary, should I work on this ? or maintain and recomment my repository ?

I've forked the repository and made the change  in here : https://flows.nodered.org/node/node-red-contrib-mongodb
It seems like there's been updates on this node since, and a merge is required so i'm not 100% sure it still works, but if it doesn't, please send me a message i'll look into it.

Merry christmas to you

knolleary commented:
Hi all,

My apologies for forgetting about this pr.

We have lots of nodes in this repository that were written in the early days of the project, before the community could write and publish their own nodes. That does mean there are nodes in here for things we don't regularly use ourselves. This does make it tricky to test/verify changes like this as I first have to setup a mongo environment to work with.

So we have a choice to make - whether we should work to make his node the "best" mongo node available, or let the community contributed nodes take that spot.

For some technologies I'm more than happy for the community to take the lead. But if there's a quick win for making our mongo node more useful then, in this instance, I'm happy to spend the time on it.

I'll see where I can get to over the Christmas holidays.

mickev36 commented:
Hi,

Thanks for your answer. No apologies needed, I understand how work goes ;-)

I understand the wish to keep one official repo for this.

If there's anything I can do to help, please let me know ! I can update the branch to remove the merge conflict if that helps.

Again, thanks for this awesome tool.

knolleary commented:
@mickev36 Updating the branch would be very welcome

mickev36 commented:
i'll look into it by the end of the week :-)

mickev36 commented:
Hello @knolleary ,

I've merged the newest changes in my repo, and fixed all the formatting issues so the diff is more clear. 
Some CI checks fail, but it's not related to my node, something might be off on the CI side.
Please let me know if you need anything else

Have a good day
